File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountApplicationFragment.java
package org.mifos.mobilebanking.ui.fragments;

/*
 * Created by saksham on 30/June/2018
 */

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountApplicationPayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountUpdatePayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsWithAssociations;
import org.mifos.mobilebanking.models.templates.savings.ProductOptions;
import org.mifos.mobilebanking.models.templates.savings.SavingsAccountTemplate;
import org.mifos.mobilebanking.presenters.SavingsAccountApplicationPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.enums.SavingsAccountState;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.SavingsAccountApplicationView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.DateHelper;
import org.mifos.mobilebanking.utils.MFDatePicker;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SavingsAccountApplicationFragment extends BaseFragment
        implements SavingsAccountApplicationView {

    @BindView(R.id.sp_product_id)
    Spinner spProductId;

    @BindView(R.id.tv_submission_date)
    TextView tvSubmissionDate;

    @BindView(R.id.tv_client_name)
    TextView tvClientName;

    @Inject
    SavingsAccountApplicationPresenter presenter;

    @Inject
    PreferencesHelper preferencesHelper;

    private View rootView;
    private SavingsAccountState state;
    private SavingsWithAssociations savingsWithAssociations;
    private ArrayAdapter<String> productIdAdapter;
    private SavingsAccountTemplate template;
    private List<ProductOptions> productOptions;
    private List<String> productIdList = new ArrayList<>();

    public static SavingsAccountApplicationFragment newInstance(
            SavingsAccountState state, SavingsWithAssociations savingsWithAssociations) {
        SavingsAccountApplicationFragment fragment = new SavingsAccountApplicationFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(Constants.SAVINGS_ACCOUNT_STATE, state);
        bundle.putParcelable(Constants.SAVINGS_ACCOUNTS, savingsWithAssociations);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            state = (SavingsAccountState) getArguments()
                    .getSerializable(Constants.SAVINGS_ACCOUNT_STATE);
            savingsWithAssociations = getArguments().getParcelable(Constants.SAVINGS_ACCOUNTS);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_savings_account_application, container,
                false);
        ButterKnife.bind(this, rootView);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);

        presenter.attachView(this);
        presenter.loadSavingsAccountApplicationTemplate(preferencesHelper.getClientId(), state);
        return rootView;
    }

    @Override
    public void showUserInterfaceSavingAccountApplication(SavingsAccountTemplate template) {
        showUserInterface(template);
    }

    @Override
    public void showSavingsAccountApplicationSuccessfully() {
        showMessage(getString(R.string.new_saving_account_created_successfully));
        getActivity().finish();
    }

    @Override
    public void showUserInterfaceSavingAccountUpdate(SavingsAccountTemplate template) {
        showUserInterface(template);
        getActivity().setTitle(getString(R.string.string_savings_account,
                getString(R.string.update)));
        spProductId.setSelection(productIdAdapter.getPosition(savingsWithAssociations
                .getSavingsProductName()));
    }

    @Override
    public void showSavingsAccountUpdateSuccessfully() {
        showMessage(getString(R.string.saving_account_updated_successfully));
        getActivity().getSupportFragmentManager().popBackStack();
    }

    public void showUserInterface(SavingsAccountTemplate template) {
        this.template = template;
        productOptions = template.getProductOptions();
        for (ProductOptions item : productOptions) {
            productIdList.add(item.getName());
        }

        tvClientName.setText(template.getClientName());

        productIdAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item,
                productIdList);
        productIdAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spProductId.setAdapter(productIdAdapter);

        tvSubmissionDate.setText(MFDatePicker.getDatePickedAsString());
    }

    public void submitSavingsAccountApplication() {
        SavingsAccountApplicationPayload payload = new SavingsAccountApplicationPayload();
        payload.setClientId(template.getClientId());
        payload.setProductId(productOptions.get(spProductId.getSelectedItemPosition()).getId());
        payload.setSubmittedOnDate(DateHelper.getSpecificFormat(DateHelper.FORMAT_dd_MMMM_yyyy,
                MFDatePicker.getDatePickedAsString()));
        presenter.submitSavingsAccountApplication(payload);
    }

    public void updateSavingAccount() {
        SavingsAccountUpdatePayload payload = new SavingsAccountUpdatePayload();
        payload.setClientId(template.getClientId());
        payload.setProductId(productOptions.get(spProductId.getSelectedItemPosition()).getId());
        presenter.updateSavingsAccount(savingsWithAssociations.getAccountNo(), payload);
    }

    @OnClick(R.id.btn_submit)
    void onSubmit() {
        if (state == SavingsAccountState.CREATE) {
            submitSavingsAccountApplication();
        } else {
            updateSavingAccount();
        }
    }

    @Override
    public void showError(String error) {
        Toast.makeText(getContext(), error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading));
    }

    @Override
    public void hideProgress() {
        hideMifosProgressDialog();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        hideProgress();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/SavingsAccountApplicationFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.databinding.FragmentSavingsAccountApplicationBinding
import org.mifos.mobile.models.accounts.savings.SavingsAccountApplicationPayload
import org.mifos.mobile.models.accounts.savings.SavingsAccountUpdatePayload
import org.mifos.mobile.models.accounts.savings.SavingsWithAssociations
import org.mifos.mobile.models.templates.savings.ProductOptions
import org.mifos.mobile.models.templates.savings.SavingsAccountTemplate
import org.mifos.mobile.ui.enums.SavingsAccountState
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.DateHelper
import org.mifos.mobile.utils.SavingsAccountUiState
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.utils.getTodayFormatted
import org.mifos.mobile.viewModels.SavingsAccountApplicationViewModel
import javax.inject.Inject

/*
* Created by saksham on 30/June/2018
*/
@AndroidEntryPoint
class SavingsAccountApplicationFragment : BaseFragment() {

    private var _binding: FragmentSavingsAccountApplicationBinding? = null
    private val binding get() = _binding!!
    private val viewModel: SavingsAccountApplicationViewModel by viewModels()


    @JvmField
    @Inject
    var preferencesHelper: PreferencesHelper? = null
    private var state: SavingsAccountState? = null
    private var savingsWithAssociations: SavingsWithAssociations? = null
    private var template: SavingsAccountTemplate? = null
    private var productOptions: List<ProductOptions>? = null
    private val productIdList: MutableList<String?> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            state = requireArguments()
                .getSerializable(Constants.SAVINGS_ACCOUNT_STATE) as SavingsAccountState
            savingsWithAssociations = arguments?.getParcelable(Constants.SAVINGS_ACCOUNTS)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentSavingsAccountApplicationBinding.inflate(inflater)
        viewModel.loadSavingsAccountApplicationTemplate(preferencesHelper?.clientId, state)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnSubmit.setOnClickListener {
            onSubmit()
        }

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.savingsAccountApplicationUiState.collect { state ->
                    when (state) {
                        SavingsAccountUiState.Loading -> showProgress()

                        SavingsAccountUiState.HideProgress -> hideProgress()

                        is SavingsAccountUiState.ShowUserInterfaceSavingAccountApplication -> {
                            hideProgress()
                            showUserInterfaceSavingAccountApplication(state.template)
                        }

                        is SavingsAccountUiState.ShowUserInterfaceSavingAccountUpdate -> {
                            hideProgress()
                            showUserInterfaceSavingAccountUpdate(state.template)
                        }

                        is SavingsAccountUiState.ErrorMessage -> {
                            hideProgress()
                            showError(state.error.message ?: "")
                        }

                        SavingsAccountUiState.SavingsAccountApplicationSuccess -> {
                            hideProgress()
                            showSavingsAccountApplicationSuccessfully()
                        }

                        SavingsAccountUiState.SavingsAccountUpdateSuccess -> {
                            hideProgress()
                            showSavingsAccountUpdateSuccessfully()
                        }

                        is SavingsAccountUiState.Initial -> {}

                        else -> throw IllegalStateException("Unexpected state : $state")
                    }

                }
            }
        }
    }

    private fun showUserInterfaceSavingAccountApplication(template: SavingsAccountTemplate?) {
        showUserInterface(template)
    }

    private fun showSavingsAccountApplicationSuccessfully() {
        showMessage(getString(R.string.new_saving_account_created_successfully))
        activity?.finish()
    }

    private fun showUserInterfaceSavingAccountUpdate(template: SavingsAccountTemplate?) {
        showUserInterface(template)
        activity?.title = getString(
            R.string.string_savings_account,
            getString(R.string.update),
        )
        binding.productIdField.setText(savingsWithAssociations?.savingsProductName!!, false)
    }

    private fun showSavingsAccountUpdateSuccessfully() {
        showMessage(getString(R.string.saving_account_updated_successfully))
        activity?.supportFragmentManager?.popBackStack()
    }

    fun showUserInterface(template: SavingsAccountTemplate?) {
        this.template = template
        productOptions = template?.productOptions
        if (productOptions != null) {
            for ((_, name) in productOptions as ArrayList<ProductOptions>) {
                productIdList.add(name)
            }
        }
        binding.tvClientName.text = template?.clientName
        binding.productIdField.setSimpleItems(productIdList.toTypedArray())
        binding.tvSubmissionDate.text = getTodayFormatted()
    }

    private fun submitSavingsAccountApplication() {
        val payload = SavingsAccountApplicationPayload()
        payload.clientId = template?.clientId

        if (productIdList.indexOf(binding.productIdField.text.toString()) != -1) {
            payload.productId = productIdList.indexOf(binding.productIdField.text.toString())
                .let { productOptions?.get(it)?.id }
        } else {
            Toaster.show(binding.root, getString(R.string.select_product_id))
            return
        }
        payload.submittedOnDate = DateHelper.getSpecificFormat(
            DateHelper.FORMAT_dd_MMMM_yyyy,
            getTodayFormatted(),
        )
        viewModel.submitSavingsAccountApplication(payload)
    }

    private fun updateSavingAccount() {
        val payload = SavingsAccountUpdatePayload()
        payload.clientId = template?.clientId?.toLong()
        payload.productId = productIdList.indexOf(binding.productIdField.text.toString())
            .let { productOptions?.get(it)?.id }?.toLong()
        viewModel.updateSavingsAccount(savingsWithAssociations?.id, payload)
    }

    private fun onSubmit() {
        if (state == SavingsAccountState.CREATE) {
            submitSavingsAccountApplication()
        } else {
            updateSavingAccount()
        }
    }

    fun showError(error: String?) {
        Toast.makeText(context, error, Toast.LENGTH_SHORT).show()
    }

    private fun showMessage(showMessage: String?) {
        Toast.makeText(context, showMessage, Toast.LENGTH_SHORT).show()
    }

    fun showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading))
    }

    fun hideProgress() {
        hideMifosProgressDialog()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onDestroy() {
        super.onDestroy()
        hideProgress()
    }

    companion object {
        @JvmStatic
        fun newInstance(
            state: SavingsAccountState?,
            savingsWithAssociations: SavingsWithAssociations?,
        ): SavingsAccountApplicationFragment {
            val fragment = SavingsAccountApplicationFragment()
            val bundle = Bundle()
            bundle.putSerializable(Constants.SAVINGS_ACCOUNT_STATE, state)
            bundle.putParcelable(Constants.SAVINGS_ACCOUNTS, savingsWithAssociations)
            fragment.arguments = bundle
            return fragment
        }
    }
}
